### ISSUE TITLE
Language Server Protocol Support

### ISSUE THREAD

**Issue by dbettin** [mawww/kakoune#1341]
Are there any plans for Kakoune to support LSP?

1. http://langserver.org/
2. https://github.com/Microsoft/language-server-protocol

**Comment by ekie** [comment]
it was discussed recently on irc and that feature would be merged. its just a matter of who will implement it.

**Comment by mawww** [comment]
Yeah, I'd be happy to have support for LSP added. I believe we already have all the needed backend features, so its just a matter or writing the support script.

**Comment by lenormf** [comment]
`clangd` alongside an LSP support script would clear up a lot of code from `extra`, as well as being potentially more maintainable.

**Comment by dbettin** [comment]
If someone wants to point me in the right direction on how to write the script, I could give it shot.

**Comment by danr** [comment]
I've started writing a monster over at https://github.com/danr/libkak

**Comment by danr** [comment]
It's actually quite usable now, give it a try!

**Comment by robertmeta** [comment]
Is LSP something that is considered "core" for Kakoune (in the editing domain, like autocomplete support)?

**Comment by doppioandante** [comment]
@robertmeta  I don't think so, AFAIK LSP is just a protocol for various editor tasks that kakoune should already support, if it can be implemented externally there's no reason to have it in core, it surely isn't at in the same tier of auto completion support.

**Comment by mawww** [comment]
LSP itself is not core, but whatever features are needed to implement it are. For example LSP requires a way to provide custom insert completion, fortunately we already have a system in place for that. It provides as well jumping to various locations in code, which should be implementatble using regular normal mode.

So I dont think we are missing any backend functionalities to add support to LSP. Parsing the protocol and converting that to the proper set of commands for Kakoune is exactly what the `%sh{...}` blocks were designed to do.

**Comment by tskinn** [comment]
Any update on this or working implementations?

**Comment by mawww** [comment]
http://github.com/ul/kak-lsp is the most advanced lsp support we have AFAIK, it works pretty well already.

**Comment by maximbaz** [comment]
The mentioned plugin works well indeed, should this issue be closed then, or you want to integrate it into kakoune itself one day?
